File: convex/rawMessages.ts
import { mutation, query } from "./_generated/server";
import { v } from "convex/values";

// Save raw message with deduplication (matches entity/raw_message.rs)
export const save = mutation({
  args: {
    externalId: v.optional(v.string()),
    groupJid: v.string(),
    groupName: v.string(), // Required in SeaORM
    senderJid: v.string(),
    senderPhone: v.optional(v.string()),
    senderName: v.optional(v.string()),
    content: v.string(),
    timestamp: v.number(),
    replyToId: v.optional(v.string()),
    replyToContent: v.optional(v.string()),
    replyToSender: v.optional(v.string()),
  },
  handler: async (ctx, args) => {
    // Check for duplicate by external ID
    if (args.externalId) {
      const existing = await ctx.db
        .query("rawMessages")
        .withIndex("by_external_id", (q) => q.eq("externalId", args.externalId))
        .first();

      if (existing) {
        // Already exists, return existing ID
        return existing._id;
      }
    }

    return await ctx.db.insert("rawMessages", {
      ...args,
      processedAt: undefined,
      error: undefined,
      createdAt: Date.now(),
    });
  },
});

// Mark as processed
export const markProcessed = mutation({
  args: {
    id: v.id("rawMessages"),
    error: v.optional(v.string()),
  },
  handler: async (ctx, args) => {
    await ctx.db.patch(args.id, {
      processedAt: Date.now(),
      error: args.error,
    });
    return await ctx.db.get(args.id);
  },
});

// Get unprocessed messages
export const getUnprocessed = query({
  args: { limit: v.optional(v.number()) },
  handler: async (ctx, args) => {
    return await ctx.db
      .query("rawMessages")
      .filter((q) => q.eq(q.field("processedAt"), undefined))
      .order("asc")
      .take(args.limit ?? 100);
  },
});

// Get by external ID
export const getByExternalId = query({
  args: { externalId: v.string() },
  handler: async (ctx, args) => {
    return await ctx.db
      .query("rawMessages")
      .withIndex("by_external_id", (q) => q.eq("externalId", args.externalId))
      .first();
  },
});

// Get recent messages for a group
export const getByGroup = query({
  args: {
    groupJid: v.string(),
    limit: v.optional(v.number()),
  },
  handler: async (ctx, args) => {
    return await ctx.db
      .query("rawMessages")
      .withIndex("by_group", (q) => q.eq("groupJid", args.groupJid))
      .order("desc")
      .take(args.limit ?? 50);
  },
});

// Delete raw messages processed before cutoff (matches SeaORM delete_before)
export const deleteBefore = mutation({
  args: { cutoff: v.number() },
  handler: async (ctx, args) => {
    const toDelete = await ctx.db
      .query("rawMessages")
      .withIndex("by_processed", (q) => q.lt("processedAt", args.cutoff))
      .collect();

    for (const doc of toDelete) {
      await ctx.db.delete(doc._id);
    }
    return toDelete.length;
  },
});

File: core/crates/db/src/convex/raw_messages.rs
//! Convex RawMessageRepository implementation

use std::sync::Arc;

use async_trait::async_trait;
use chrono::{DateTime, Utc};

use super::client::ConvexClient;
use crate::Result;
use crate::convex_args;
use crate::traits::{RawMessageModel, RawMessageRepository};

/// Convex-backed raw message repository
pub struct ConvexRawMessageRepo {
    client: Arc<ConvexClient>,
}

impl ConvexRawMessageRepo {
    pub fn new(client: Arc<ConvexClient>) -> Self {
        Self { client }
    }
}

#[async_trait]
impl RawMessageRepository for ConvexRawMessageRepo {
    async fn save(&self, message: &RawMessageModel) -> Result<RawMessageModel> {
        // RawMessageModel: id, external_id, group_jid, group_name, sender_jid, sender_phone, sender_name,
        // content, timestamp, processed_at, error, reply_to_id, reply_to_content, reply_to_sender, created_at
        let id: String = self
            .client
            .mutation(
                "rawMessages:save",
                convex_args! {
                    "externalId" => message.external_id.as_ref(),
                    "groupJid" => &message.group_jid,
                    "groupName" => &message.group_name,
                    "senderJid" => &message.sender_jid,
                    "senderPhone" => message.sender_phone.as_ref(),
                    "senderName" => message.sender_name.as_ref(),
                    "content" => &message.content,
                    "timestamp" => message.timestamp.timestamp_millis(),
                    "replyToId" => message.reply_to_id.as_ref(),
                    "replyToContent" => message.reply_to_content.as_ref(),
                    "replyToSender" => message.reply_to_sender.as_ref()
                },
            )
            .await?;

        let mut saved = message.clone();
        saved.id = id;
        Ok(saved)
    }

    async fn get_by_id(&self, id: &str) -> Result<Option<RawMessageModel>> {
        self.client
            .query(
                "rawMessages:getByExternalId",
                convex_args! { "externalId" => id },
            )
            .await
    }

    async fn get_unprocessed(&self, limit: i64) -> Result<Vec<RawMessageModel>> {
        self.client
            .query(
                "rawMessages:getUnprocessed",
                convex_args! { "limit" => limit },
            )
            .await
    }

    async fn mark_processed(&self, id: &str, error: Option<&str>) -> Result<RawMessageModel> {
        self.client
            .mutation(
                "rawMessages:markProcessed",
                convex_args! {
                    "id" => id,
                    "error" => error
                },
            )
            .await
    }

    async fn delete_before(&self, cutoff: &DateTime<Utc>) -> Result<u64> {
        self.client
            .mutation(
                "rawMessages:deleteBefore",
                convex_args! { "cutoff" => cutoff.timestamp_millis() },
            )
            .await
    }
}
